dbms/src/Storages/ReadInOrderOptimizer.cpp

```diff
@@ -0,0 +1,111 @@
+#include <Storages/ReadInOrderOptimizer.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+
+ReadInOrderOptimizer::ReadInOrderOptimizer(
+    const ManyExpressionActions & elements_actions_,
+    const SortDescription & required_sort_description_,
+    const SyntaxAnalyzerResultPtr & syntax_result)
+    : elements_actions(elements_actions_)
+    , required_sort_description(required_sort_description_)
+{
+    if (elements_actions.size() != required_sort_description.size())
+        throw Exception("Sizes of sort description and actions are mismatched", ErrorCodes::LOGICAL_ERROR);
+
+    /// Do not analyze ARRAY JOIN result columns.
+    /// TODO: forbid more columns for analyzing.
+    for (const auto & elem : syntax_result->array_join_result_to_source)
+        forbidden_columns.insert(elem.first);
+}
+
+InputSortingInfoPtr ReadInOrderOptimizer::analyze(const StoragePtr & storage) const
```

Comment:
What does this function do, does it apply a projection (in relational sense) to a merge tree sorting key? E.g. you have some expressions derived from primary columns, you have the input sorting key in terms of primary columns, and you try to pull this key through the expressions and see what it becomes? Maybe we can think of a better name for the function based on this.